---
sort: 1
---
## dadaist2
**dadaist2** - a shell wrapper for DADA2, to detect representative sequences and generate
a feature table starting from Illumina Paired End reads.

## Author
Andrea Telatin <andrea.telatin@quadram.ac.uk>

## Synopsis
    dadaist2 [options] -i INPUT_DIR -o OUTPUT_DIR

## Parameters
## Main Parameters

- _-i_, _--input-directory_ DIRECTORY

    Directory containing the paired end files in FASTQ format, gzipped or not.

- _-o_, _--output-directory_ DIRECTORY

    Output directory (will be created).

- _-m_, _--metadata_ FILE

    Metadata file in TSV format, first column must match sample IDs. If not supplied
    a template will be autogenerated using `makeSampleSheet`.

- _-d_, _--database_ DATABASE

    Reference database in gzipped FASTA format, specify 'skip' not to assign
    taxonomy (default: skip)

- _-t_, _--threads_ INT

    Number of threads (default: 2)

- _--primers_ FOR:REV

    Strip primers with amplicheck, otherwise trim the primer regions
    with fastp.

- _-j_, _--just-concat_

    Do not try merging paired end reads, just concatenate.

## Metabarcoding processing

- _-q_, _--min-qual_ FLOAT

    Minimum average quality for DADA2 truncation (default: 28)

- _--no-trunc_

    Do not truncate reads (required for non-overlapping amplicons, like ITS)

- _--maxee1_, and _--maxee2_ FLOAT

    Maximum Expected Errors in R1 and R2, respectively (default: 1.0 and 1.5)

- _s1_, _--trim-primer-for_ INT

    Trim primer from R1 read specifying the number of bases. Similarly
    use `-s2` (`--trim-primer-rev`) to remove the front bases from the
    reverse pair (R2). Default: 20 bases each side.

- _-r_, _--save-rds_

    Save a copy of the RDS file (default: off)

- _-1_, _--for-tag_ STRING \[and _-2_, _--rev-tag_\]

    String identifying the forward (and reverse) pairs. Default are \_R1 and \_R2.

## Other parameters

- _-s_, _--id-separator_ STRING

    String used to separate the "sample name" from the filename. Default "\_".

- _-p_, _--prefix_ STRING

    Prefix for the output FASTA file, if "MD5" is specified, the sequence MD5 hash
    will be used instead. Default is "ASV".

- _-l_, _--log-file_ FILE

    Filename for the program log.

- _--popup_

    Display popup notifications (tested on MacOS and Ubuntu)

## Source code and documentation
The program is freely available at [https://quadram-institute-bioscience.github.io/dadaist2](https://quadram-institute-bioscience.github.io/dadaist2)
released under the MIT licence. The website contains further DOCUMENTATION.
