---
sort: 4
permalink: /usage
---

# Usage guide

## Input files

* A directory with Paired-End FASTQ files (gzipped or not)
* (optional) a metadata file

The input directory should contain paired-end reads as follows (with identifiable
strand tags, like *_R1* and *_R2*, and a sample ID separator like *_*):
```
input_dir/Sample1_R1.fastq.gz
input_dir/Sample1_R2.fastq.gz
input_dir/Sample2_R1.fastq.gz
input_dir/Sample2_R2.fastq.gz
input_dir/Sample3_R1.fastq.gz
input_dir/Sample3_R2.fastq.gz
```

## Running dadaist2

Without taxonomical assignment:
```
dadaist2 -i input_directory -o dadaist2_output --threads 8
```

With a reference database (in FASTQ gzipped format):
```
dadaist2 -d refs/database.fq.gz -i input_directory -o dadaist2_output --threads 8
```

With a specific metadata file:
```
dadaist2 -m metadata.tsv -d refs/database.fq.gz -i input_directory -o dadaist2_output --threads 8
```


## Output

The output directory will contain:
* **dada2_stats.tsv**, a table with number of raw, filtered, denoised and merged sequences.
* **rep-seqs.fasta**, FASTA file with the _representative sequences_ identified in the experiment.
* **feature-table.tsv**, _feature table_ with the raw counts of each representative sequence in each sample.
* **metadata.tsv**, a copy of the metadata file (or a blank metadata file autogenerated, if not provided).
* **qc**, directory with quality plots and JSON file of the quality profile of each sample (produced by _fastp_).
* **rep-seqs.msa**, multiple sequence alignment of the representative sequences.
* **rep-seqs.tree**, tree generated from the multiple sequence alignment.
* **dadaist.log**, program log
* **taxonomy.txt** (if a reference is provided), classification of each representative sequence.
